PortalSwift/Sources/PortalSwift/Provider/PortalProvider.swift

```diff
@@ -0,0 +1,702 @@
+  }
+
+  /// Registers a callback for an event.
+  /// - Parameters:
+  ///   - event: The event to register a callback.
+  ///   - callback: The function to be invoked whenever the event fires.
+  /// - Returns: The Portal Provider instance.
+  public func on(event: Events.RawValue, callback: @escaping (_ data: Any) -> Void) -> PortalProvider {
+    if self.events[event] == nil {
+      self.events[event] = []
+    }
+
+    self.events[event]?.append(RegisteredEventHandler(
+      handler: callback,
+      once: false
+    ))
+
+    return self
+  }
+
+  /// Registers a callback for an event. Deletes the registered callback after it's fired once.
+  /// - Parameters:
+  ///   - event: The event to register a callback.
+  ///   - callback: The function to be invoked whenever the event fires.
+  /// - Returns: The Portal Provider instance.
+  public func once(
+    event: Events.RawValue,
+    callback: @escaping (_ data: Any) throws -> Void
+  ) -> PortalProvider {
+    if self.events[event] == nil {
+      self.events[event] = []
+    }
+
+    self.events[event]?.append(RegisteredEventHandler(
+      handler: callback,
+      once: true
+    ))
+
+    return self
+  }
+
+  /// Removes the callback for the specified event.
+  /// - Parameters:
+  ///   - event: A specific event from the list of Events.
+  /// - Returns: An instance of Portal Provider.
+  public func removeListener(
+    event: Events.RawValue
+  ) -> PortalProvider {
+    if self.events[event] == nil {
+      self.logger.info(String(format: "[Portal] Could not find any bindings for event '%@'. Ignoring...", event))
+    }
+
+    self.events[event] = nil
+
+    return self
+  }
+
+  /// Makes a request.
+  /// - Parameters:
+  ///   - chainId: A CAIP-2 Blockchain ID associated with the request.
+  ///   - withMethod: A member of the PortalRequestMethod enum
+  ///   - andParams: An array of parameters for the request (either RPC parameters or a transaction if signing)
+  /// - Returns: PortalProviderResult
+  public func request(
+    _ chainId: String,
+    withMethod: PortalRequestMethod,
+    andParams: [AnyEncodable]? = [],
+    connect: PortalConnect? = nil
+  ) async throws -> PortalProviderResult {
+    let blockchain = try PortalBlockchain(fromChainId: chainId)
+    guard blockchain.isMethodSupported(withMethod) else {
+      throw PortalProviderError.unsupportedRequestMethod(withMethod.rawValue)
+    }
+
+    let id = UUID().uuidString
+
+    // This switch is here to handle methods that should be
+    // resolved by the provider directly before passing the
+    // request on to RPC or the signer.
+    //
+    // The default behavior is to use the `PortalBlockchain`
+    // instance to determine if the method should be signed
+    // or not.
+    switch withMethod {
+    case .eth_accounts, .eth_requestAccounts:
+      let address = try await keychain.getAddress(chainId)
+      return PortalProviderResult(id: id, result: [address])
+    case .wallet_switchEthereumChain:
```

Comment:
These events will only ever come from dApps via `PortalConnect` or the `InjectedProvider`. Those classes are now maintaining their own `chainId` but the event is still emitted. The contract for this RPC method requires that `null` be returned for a successful call, so that's still here until we fully deprecate all of the `setChainId` stuff from the Provider itself (after customers have fully migrated to chain-agnosticism)